go_sequencer/common/account.go

```diff
@@ -1,67 +1,195 @@
 	// from the User Idxs can be
-	IdxUserThreshold = Idx(UserThreshold)
+	IdxUserThreshold = AccountIdx(UserThreshold)
 )
 
-// Bytes returns a byte array representing the Idx
-func (idx Idx) Bytes() ([6]byte, error) {
-	if idx > maxIdxValue {
-		return [6]byte{}, Wrap(ErrIdxOverflow)
+var (
+	// FFAddr is used to check if an ethereum address is 0xff..ff
+	FFAddr = ethCommon.HexToAddress("0xffffffffffffffffffffffffffffffffffffffff")
+	// EmptyAddr is used to check if an ethereum address is 0
+	EmptyAddr = ethCommon.HexToAddress("0x0000000000000000000000000000000000000000")
+)
+
+// Bytes returns a byte array representing the accountIdx
+func (idx AccountIdx) Bytes() ([NLevelsAsBytes]byte, error) {
+	if idx > maxAccountIdxValue {
+		return [NLevelsAsBytes]byte{}, Wrap(ErrIdxOverflow)
 	}
 	var idxBytes [8]byte
 	binary.BigEndian.PutUint64(idxBytes[:], uint64(idx))
-	var b [6]byte
-	copy(b[:], idxBytes[2:])
+	var b [NLevelsAsBytes]byte
+	copy(b[:], idxBytes[8-NLevelsAsBytes:])
 	return b, nil
 }
 
-// IdxFromBytes returns Idx from a byte array
-func IdxFromBytes(b []byte) (Idx, error) {
+// AccountIdxFromBytes returns Idx from a byte array
+func AccountIdxFromBytes(b []byte) (AccountIdx, error) {
 	if len(b) != IdxBytesLen {
 		return 0, Wrap(fmt.Errorf("can not parse Idx, bytes len %d, expected %d",
 			len(b), IdxBytesLen))
 	}
 	var idxBytes [8]byte
-	copy(idxBytes[2:], b[:])
+	copy(idxBytes[8-NLevelsAsBytes:], b[:])
 	idx := binary.BigEndian.Uint64(idxBytes[:])
-	return Idx(idx), nil
+	return AccountIdx(idx), nil
+}
+
+// BigInt returns a *big.Int representing the Idx
+func (idx AccountIdx) BigInt() *big.Int {
+	return big.NewInt(int64(idx))
+}
+
+// Bytes returns the bytes representing the Account, in a way that each BigInt
+// is represented by 32 bytes, in spite of the BigInt could be represented in
+// less bytes (due a small big.Int), so in this way each BigInt is always 32
+// bytes and can be automatically parsed from a byte array.
+func (a *Account) Bytes() ([32 * NAccountLeafElems]byte, error) {
+	var b [32 * NAccountLeafElems]byte
+
+	if a.Nonce > nonce.MaxNonceValue {
+		return b, Wrap(fmt.Errorf("%s Nonce", ErrNumOverflow))
+	}
+	if len(a.Balance.Bytes()) > maxBalanceBytes {
+		return b, Wrap(fmt.Errorf("%s Balance", ErrNumOverflow))
+	}
+
+	nonceBytes, err := a.Nonce.Bytes()
+	if err != nil {
+		return b, Wrap(err)
+	}
+
+	copy(b[23:28], nonceBytes[:])
+
+	pkSign, pkY := babyjub.UnpackSignY(a.BJJ)
+	if pkSign {
+		b[22] = 1
+	}
+	balanceBytes := a.Balance.Bytes()
+	copy(b[64-len(balanceBytes):64], balanceBytes)
+	// Check if there is possibility of finite field overflow
+	ayBytes := pkY.Bytes()
+	if len(ayBytes) == 32 { //nolint:gomnd
+		ayBytes[0] = ayBytes[0] & 0x3f //nolint:gomnd
+		pkY = big.NewInt(0).SetBytes(ayBytes)
+	}
+	finiteFieldMod, ok := big.NewInt(0).SetString("21888242871839275222246405745257275088548364400416034343698204186575808495617", 10) //nolint:gomnd
```

Comment:
In elliptic curve, pkY should be inside the finiteField so that we can use it for further purpose.
And regarding the size of pkY, it is too long to store, so we need to mod pkY by finiteFieldMod, which is 32 bytes.